default/AI/ProductionAI.py

```diff
@@ -92,151 +142,186 @@ def getBestShipInfo(priority, loc=None):
-        design = fo.getShipDesign(design_id)
-        valid_locs = []
-        for pid in planet_ids:
-            if pid is None:
-                continue
-            if design.productionLocationForEmpire(empire_id, pid):
-                valid_locs.append(pid)
-        if valid_locs:
-            return design_id, design, valid_locs
-    return None, None, None  # must be missing a Shipyard or other orbital (or missing tech)
+
+    if priority in design_cache:  # use new framework
+        bestDesigns = design_cache[priority]
+        if not bestDesigns:
+            return None, None, None
+        topRating, topID = bestDesigns[0][0], bestDesigns[0][2]
+        validLocs = [item[1] for item in bestDesigns if item[0] == topRating and item[2] == topID]
+        return topID, fo.getShipDesign(topID), validLocs
+    else:  # use old framework
+        empire = fo.getEmpire()
+        empire_id = empire.empireID
+        these_design_ids = []
+        design_name_bases = shipTypeMap.get(priority, ["nomatch"])
+        for base_name in design_name_bases:
+            these_design_ids.extend([(design_name_bases[base_name] + fo.getShipDesign(design).name(False), design)
+                                     for design in empire.availableShipDesigns
+                                     if base_name in fo.getShipDesign(design).name(False)])
+        if not these_design_ids:
+            return None, None, None  # must be missing a Shipyard (or checking for outpost ship but missing tech)
+        # ships = [(fo.getShipDesign(design).name(False), design) for design in these_design_ids]
+
+        for _, design_id in sorted(these_design_ids, reverse=True):
+            design = fo.getShipDesign(design_id)
+            valid_locs = []
+            for pid in planet_ids:
+                if pid is None:
+                    continue
+                if design.productionLocationForEmpire(empire_id, pid):
+                    valid_locs.append(pid)
+            if valid_locs:
+                return design_id, design, valid_locs
+        return None, None, None  # must be missing a Shipyard or other orbital (or missing tech)
 
 
-def getBestShipRatings(loc=None, verbose = False):
+def getBestShipRatings(loc=None, verbose=False):
     """returns list of [partition, pid, designID, design] sublists, currently only for military ships"""
     priority = EnumsAI.AIPriorityType.PRIORITY_PRODUCTION_MILITARY
-    empire = fo.getEmpire()
-    empireID = empire.empireID
-    capitolID = PlanetUtilsAI.get_capital()
-    cur_turn = fo.currentTurn()
     if loc is None:
         planetIDs = ColonisationAI.empire_shipyards
     elif isinstance(loc, list):
-        planetIDs=set(loc).intersection(ColonisationAI.empire_shipyards)
+        planetIDs = set(loc).intersection(ColonisationAI.empire_shipyards)
     elif isinstance(loc, int):
         if loc in ColonisationAI.empire_shipyards:
-            planetIDs=[loc]
+            planetIDs = [loc]
```

Comment:
I have to admit I actually find the option to pass a single planet as ```pid``` instead of ```[pid]``` intuitive and more user-friendly.

The functionality that a list of planets can be handled should remain as option, I think.